timemachine/md/smc.py

```diff
@@ -301,6 +301,46 @@ def multinomial_resample(log_weights):
     return indices, avg_log_weights
 
 
+def stratified_resample(log_weights):
+    """
+    Split cummulative sum of weights into N interval
+    and pick one particle per subinterval.
+
+    References
+    ----------
+    * [Douc, Cappé, Moulines, 2005]
+        Comparison of Resampling Schemes for Particle Filtering
+        https://arxiv.org/abs/cs/0507025
+
+    * Code based on:
+        https://github.com/rlabbe/filterpy/blob/6cc052bea5a806d3531b59c820530c42b67a910c/filterpy/monte_carlo/resampling.py
+        [MIT license]
+    """
+    weights = np.exp(log_weights - logsumexp(log_weights))
+    assert np.isclose(np.sum(weights), 1.0)
+    n = len(log_weights)
+
+    # generate n subintervals that are on average 1/n apart
+    subintervals = (np.random.random(n) + np.arange(n)) / n
+
+    # init
+    indices = np.zeros(n, dtype=int)
+    cumulative_sum = np.cumsum(weights)
+
+    i, j = 0, 0
+    while i < n:
+        if subintervals[i] < cumulative_sum[j]:
+            indices[i] = j
+            i += 1
+        else:
+            j += 1
+
+    # new weights
+    avg_log_weights = logsumexp(log_weights - np.log(n)) * np.ones(n)
+
+    return indices, avg_log_weights
```

Comment:
opinion: another preferred MIT licensed source of reference SMC implementations in python is the `nchopin/particles` repository, https://github.com/nchopin/particles/blob/a760322e3e0f5e0888334373d784f52aec917b12/particles/resampling.py#L599-L603 , accompanying his book. (Which coincides with this implementation, I think, for the case `len(weights) == M`)